Fix serialization

DemiMarie opened the pull request.

DemiMarie committed:
Working serialization

Bignum serialization now works, for both positive and negative bignums.

DemiMarie committed:
Implement main program and proof-of-time CLI

This implements proving in the CLI.  The proofs have been tested against
the reference `pot` tool.  Verification does not work, however.

DemiMarie committed:
Fix bugs in multiplication

This fixes a bug in multiplication that caused a crash (assertion
failure) when the number of iterations was ≥ 130.  The code can now
verify its own proofs for t ≤ 128, and matches the reference
implementation for t ≤ ~500.

DemiMarie committed:
Add and fix tests.

Some of the tests were bogus, as they passed invalid parameters to
functions whose behavior on such parameters is unspecified.
Furthermore, add a large number of tests for multiplication and
squaring.

DemiMarie committed:
Fix proof generation

The proof now works correctly in many more cases.  I (Demi) still need
to do additional stress tests, though.

DemiMarie committed:
Run `cargo fmt`

DemiMarie committed:
Replace gen_primes.py with a build script

I checked that both produce the same output (modulo formatting).  This
change avoids having to either have a build-time dependency on Python or
to include a large generated blob in the repository.

DemiMarie committed:
Replace an `assert!` with a `debug_assert!`

This should make things slightly faster in release builds.

DemiMarie committed:
Make build-script generated constants private

DemiMarie committed:
Fix typo in comment in create_discriminant.rs

DemiMarie committed:
Remove outdated FIXME

DemiMarie committed:
Fix length calculation in deserialization

Previously, all proofs that had a proof blob were rejected at the
deserialization step.

The code is now able to verify some of its own proofs for long numbers
of iterations.

DemiMarie committed:
Make some code more generic over the bignum type

afck commented on check.sh:
```diff
@@ -0,0 +1,3 @@
+#!/bin/sh --
```
Good idea!
We should (later, in a separate PR) rename this to `ci.sh`, enable Travis and make this run for each PR, just like in [hbbft](https://github.com/poanetwork/hbbft/) and [threshold_crypto](https://github.com/poanetwork/threshold_crypto).

DemiMarie replied:
I agree, though it will need various test cases to run.

afck commented on vdf-cli/src/main.rs:
```diff
@@ -14,47 +14,172 @@
+    for i in obj.chars() {
+        match i {
+            '0'...'9' | 'a'...'f' | 'A'...'F' => (),
+            _ => drop(hex::decode(&obj).map_err(|x| format!("{}", x))),
```
Is it really necessary to iterate through the chars here? Wouldn't `hex::decode` fail anyway otherwise?

DemiMarie replied:
The way I wrote the code speeds up the happy path, but that might not be worth it.

afck replied:
But this is nothing that will happen inside any inner loop, is it? It's just for checking the command line argument? I'd prefer avoiding any premature optimization there, and instead keep the code as simple (and auditable!) as possible.

DemiMarie replied:
I agree.

DemiMarie replied:
I suspect that the double-allocation requirement is either a bug in clap, or a problem with how I am using it.

afck commented on vdf-cli/src/main.rs:
```diff
@@ -14,47 +14,172 @@
-    let iterations = value_t!(_matches, "NUM_ITERATIONS", u64).unwrap();
-    let discriminant = Mpz::from_str_radix(_matches.value_of("DISCRIMINANT").unwrap(), 0).unwrap();
-    println!("{}", vdf::do_compute(discriminant, iterations))
+    if let Some(_matches) = matches.subcommand_matches("compute") {
```
Since these are quite a few subcommands, `match matches.subcommand()` might be cleaner than lots of `else if`s.

afck commented on vdf/build.rs:
```diff
@@ -0,0 +1,101 @@
+const PRIMES_LEN: usize = 6541;
+const RESIDUES_LEN: usize = 5760;
+const SIEVE_INFO_LEN: usize = PRIMES_LEN - 5;
+fn odd_primes_below_n(n: usize) -> Vec<usize> {
```
I'd put a new line between `use`s, `const`s and `fn`s.
(In fact, I'd use `rustfmt`—but that's optional, of course!)

DemiMarie replied:
Done, at least in this file.

afck commented on vdf/build.rs:
```diff
@@ -0,0 +1,101 @@
+use std::u16;
+const PRIMES_LEN: usize = 6541;
+const RESIDUES_LEN: usize = 5760;
+const SIEVE_INFO_LEN: usize = PRIMES_LEN - 5;
```
I have no idea where these numbers come from; it would be helpful to have some comments on them, even if they are short and just point the reader in the right direction.

DemiMarie replied:
Done.

afck commented on vdf/build.rs:
```diff
@@ -0,0 +1,101 @@
+fn odd_primes_below_n(n: usize) -> Vec<usize> {
+    let mut sieve = Vec::with_capacity(n >> 1);
+    sieve.resize(n >> 1, true);
+    let mut q = (n as f64).powf(0.5) as usize;
```
Would it be safer to use a crate like [integral_square_root](https://crates.io/crates/integral_square_root) or [integer-sqrt-rs](https://github.com/paritytech/integer-sqrt-rs) for this? (Or others—not sure which ones are good.) I don't know whether it's always guaranteed otherwise that `q` is the largest number with `q * q <= n`. But maybe it actually is?

DemiMarie replied:
That is what the subsequent `assert!`s are for: they verify that `q` *is* the actual square root.  This function only needs to work for the specific number (65536) that it is called with.  Therefore, it is okay if it misbehaves for other inputs.

afck commented on vdf/build.rs:
```diff
@@ -0,0 +1,101 @@
+            res.push(2 * i + 1);
+        }
+    }
+    assert_eq!(res.len(), PRIMES_LEN);
```
But that depends on `n`, I guess? Maybe `n` shouldn't be a parameter of this function at all then.

DemiMarie replied:
It does.  As I wrote above, this function will only be called with a single specific input.

afck replied:
But then why have an input at all, instead of making that specific input a constant, too?

DemiMarie replied:
Good point.  Fixed.

afck commented on vdf/build.rs:
```diff
@@ -0,0 +1,101 @@
+}
+
+// Probably horribly slow.  I don’t care.
+fn mod_exponentiation(q: usize, r: usize, s: usize) -> usize {
```
Please document the arguments and/or give them more descriptive names.

DemiMarie replied:
Done.

afck commented on vdf/build.rs:
```diff
@@ -0,0 +1,101 @@
+}
+
+const M: usize = 8 * 3 * 5 * 7 * 11 * 13;
+fn generate(f: &mut dyn Write) {
```
Please also give these a more descriptive name and doc comment.

DemiMarie replied:
I am not sure what to put for `M`, since it is just a number.

afck replied:
Then maybe just a comment. Something like:

```markdown
A product of small prime numbers. We precompute the numbers between `1` and `M` that are
coprime to `M`. Any number whose residue modulo `M` is one of these is not divisible by any
of the prime factors of `M`. This speeds up the generation of random prime numbers.
```

afck commented on vdf/src/classgroup.rs:
```diff
@@ -24,7 +24,8 @@ pub enum InvalidDiscriminant<T> {
 #[allow(type_alias_bounds)] // false positive
 pub type Result<T: ClassGroup> = std::result::Result<T, InvalidDiscriminant<T::BigNum>>;
```
I think this should be:

```rust
pub type Result<T> = std::result::Result<T, InvalidDiscriminant<<T as ClassGroup>::BigNum>>;
```

Then the above `#[allow(type_alias_bounds)]` shouldn't be required.

DemiMarie replied:
Done.

afck commented on vdf/src/classgroup.rs:
```diff
@@ -52,11 +55,15 @@ where
+    fn serialize(&self, buf: &mut [u8]) -> std::result::Result<(), usize>;
+
+    /// Deserializes a bignum from raw, big-endian bytes.  The bytes **must not**
+    /// be interpreted as 2’s complement.
```
It would be better so say what they _should_ be implemented as. (First bit is the sign, the rest is an unsigned integer?)

DemiMarie replied:
Done.

afck commented on vdf/src/classgroup.rs:
```diff
@@ -24,7 +24,8 @@ pub enum InvalidDiscriminant<T> {
 pub type Result<T: ClassGroup> = std::result::Result<T, InvalidDiscriminant<T::BigNum>>;
 
-pub trait ClassGroup: Sized + Clone + for<'a> MulAssign<&'a Self> + PartialEq
+pub trait ClassGroup:
```
How is `generate_for_discriminant` different from `identity_for_discriminant`? Does it generate a random element? Please add that to the method's documentation.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+    while blob.len() < byte_count {
+        let mut hasher = Sha256::new();
+        hasher.input(seed);
+        let extra_bits: [u8; 2] = unsafe { mem::transmute(extra.to_be()) };
```
Would this be possible without unsafe code, with a crate like [byteorder](https://crates.io/crates/byteorder)?

DemiMarie replied:
Probably simpler to just write it out by hand as

```rust
let extra_bits = [((extra & 0xFF00) >> 8) as u8, (extra & 0xFF) as u8];
```

DemiMarie replied:
Done.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+
+pub fn create_discriminant(seed: &[u8], length: u16) -> Mpz {
+    let extra: u8 = (length as u8) & 7;
+    let entropy_bytes = ((length >> 3) + if extra == 0 { 2 } else { 3 }) as usize;
```
Please add some documentation.
If I'm reading this correctly, `length` is in bits, and somehow `entropy_bytes` is either the length in bytes plus 2 (if the length is a multiple of `8`) or plus 3 (if it isn't)? Why?

Also, I wonder whether division would be more readable than all those right shifts; I'm sure the compiler is smart enough to compile them to the same code.

Maybe this would be cleaner?

```rust
let entropy_bytes = (((length + 7) / 8) + 2) as usize;
```

DemiMarie replied:
Fixed (though still with the right shifts).  You are almost certainly correct about optimization, though.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+
+//! Creation of discriminants.
+//!
+//! The `pot` tool does not accept a discriminant as a command-line argument.
```
it would be really helpful to add a permalink to the `pot` code here, maybe even to the relevant file.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+    assert_eq!(entropy.len(), entropy_bytes);
+    let residue = {
+        let last_2 = &mut entropy[entropy_bytes - 2..];
+        let numerator = (usize::from(last_2[0]) << 8) + usize::from(last_2[1]);
```
I see… so it's two bytes for the numerator, and the `3` above is just to accommodate for the additional bits of the rest…?

DemiMarie replied:
Yes.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+use std::mem;
+use std::u16;
+
+fn entropy_from_seed(seed: &[u8], byte_count: usize) -> Vec<u8> {
```
This doesn't really add any entropy in addition to the seed's, I think. Maybe `random_bytes_from_seed` would be more appropriate?

DemiMarie replied:
Fixed.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+    {
+        mpz_add_ui_self(&mut n, residue);
+    }
```
Do these braces do anything? It doesn't look like anything would be dropped or any borrow would end earlier due to them?

DemiMarie replied:
No, they don’t.  Fixed.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+        n -= r;
+    }
+    {
+        mpz_add_ui_self(&mut n, residue);
```
OK, so we truncated `n` from the right to its desired bit length, set the last bit to make it an odd number, then made it divisible by `M` and added `residue`… (we replaced its residue modulo `M` by some constant)? Is this all to make it more likely to be prime or is there more to it?

DemiMarie replied:
From what I understand, it makes the distribution more uniform.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+            }
+        }
+        n += (u64::from(M) * (1 << 16)) as u64
+    }
```
Wasn't there another prime sieve implementation above? Could that be deduplicated?

DemiMarie replied:
I am not certain what this sieve is for.  It doesn’t look like a prime sieve.

DemiMarie replied:
It actually is a prime sieve, just not the standard one.  I don’t think it can be deduplicated, though.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -0,0 +1,112 @@
+    }
+    loop {
+        let mut sieve = Vec::with_capacity(1 << 16);
+        sieve.resize(1 << 16, false);
```
Why not just `let mut sieve = vec![false; 1 << 16]`?

DemiMarie replied:
Done.

afck commented on vdf/src/gmp_classgroup/congruence.rs:
```diff
@@ -35,6 +35,7 @@ impl Default for CongruenceContext {
         }
     }
 }
+
```
Please also document `CongruenceContext`—these are essentially temporary values for a computation, and just exist to avoid repeated allocation, right? (Also the methods.)

DemiMarie replied:
Done.

afck commented on vdf/src/gmp_classgroup/ffi.rs:
```diff
@@ -103,19 +123,41 @@ pub fn mpz_fdiv_q_ui_self(rop: &mut Mpz, op: c_ulong) -> c_ulong {
+        }
+        obj
+    }
+    let is_negative = buf.first().expect("Cannot deserialize an empty buffer") >> 7;
```
This would be more descriptive if you added the `== 1` (or `!= 0`) to this line, so `is_negative` would be a `bool`.

DemiMarie replied:
Fixed.

afck commented on vdf/src/gmp_classgroup/ffi.rs:
```diff
@@ -149,34 +191,117 @@ pub fn mpz_fdiv_q_self(rop: &mut Mpz, op: &Mpz) {
 }
 
-pub fn export_obj(obj: &Mpz, v: &mut [u8]) -> Result<(), ()> {
+pub fn export_obj(obj: &Mpz, v: &mut [u8]) -> Result<(), usize> {
```
Please also document what this does; especially the meaning of errors.

DemiMarie replied:
Done.

afck commented on vdf/src/gmp_classgroup/mod.rs:
```diff
@@ -69,46 +69,166 @@ impl fmt::Display for GmpClassGroup {
     fn inner_reduce(&mut self, ctx: &mut Ctx) {
+        self.assert_valid();
         self.inner_normalize(ctx);
+        self.assert_valid();
```
That seems redundant, now that the assertions are also inside `inner_normalize`?

DemiMarie replied:
I removed this and other redundant assertions.

afck commented on vdf/src/gmp_classgroup/mod.rs:
```diff
@@ -200,13 +336,19 @@ impl Default for GmpClassGroup {
 }
 
 impl MulAssign<Self> for GmpClassGroup {
+    #[cfg_attr(not(debug_assertions), inline(always))]
```
Maybe the number of `Mul[Assign]` impls could be halved by using `Borrow`—that should cover both the reference and value cases:

```rust
impl<B: Borrow<GmpClassGroup>> MulAssign<B> for GmpClassGroup { … }
```

afck reviewed (commented):
Great work!
I'm only partway through the review so far. I added some comments.

afck commented on vdf/src/proof_of_time.rs:
```diff
@@ -12,14 +12,13 @@
 fn generate_r_value<T>(x: &T, y: &T, sqrt_mu: &T, int_size_bits: usize) -> Result<T::BigNum, ()>
 where
-    T: ClassGroup<BigNum = super::gmp_classgroup::Mpz>,
+    T: ClassGroup,
```
Should the other two bounds below be part of the `ClassGroup` trait?

afck replied:
I see, they're already there in the `where` clause… I guess that doesn't work then. :confused:

DemiMarie replied:
Correct, sadly.

afck commented on vdf/src/proof_pietrzak.rs:
```diff
@@ -208,3 +216,36 @@ where
+    }
+    #[test]
+    fn check_cache_indeces() {
+        assert_eq!(cache_indeces_for_count(66)[..], [33, 66]);
```
Typo: "indices" or "indexes".

DemiMarie replied:
Fixed.

afck commented on vdf/src/proof_of_time.rs:
```diff
@@ -102,18 +109,27 @@ where
-        v.push(0)
-    }
-    y.serialize(&mut v[0..element_length])?;
+    v.resize(proof_len_in_bytes, 0);
```
Would `let mut v = vec![0; proof_len_in_bytes]` do the same?

DemiMarie replied:
It would, thank you!

DemiMarie replied:
Done.

afck commented on vdf/src/proof_pietrzak.rs:
```diff
@@ -12,13 +12,14 @@
-    let x: f64 = (t as f64) / 16.;
-    let w = x.log2() - x.log2().log2() + 0.25;
-    (w / 2.).round() as _
+    let x: f64 = ((t as f64) / 16.) * 2.0f64.ln();
```
Please document what this approximates.

DemiMarie replied:
Done

afck commented on vdf/tests/check_multiply.rs:
```diff
@@ -0,0 +1,75 @@
+    let manifest_path = std::env::var("CARGO_MANIFEST_DIR").expect("cargo should have set this");
+    let mut path = PathBuf::from(&manifest_path);
+    path.push("tests/multiply.txt");
+    let mut f = BufReader::new(File::open(path).expect("test malfunction"));
```
`"test file missing"`?

DemiMarie replied:
Done (with a slightly different message).

afck commented on vdf/tests/check_multiply.rs:
```diff
@@ -0,0 +1,75 @@
+    let mut f = BufReader::new(File::open(path).expect("test malfunction"));
+    let mut buffer = String::new();
+    loop {
+        let bytes_read = f.read_line(&mut buffer).expect("test malfunction");
```
`"could not read from test file"`?

DemiMarie replied:
Done.

afck reviewed (commented):
That's the rest of my first (rather superficial) pass through the code.
I haven't really reviewed the multiplication and normalization implementations in detail, and some other theory-heavy parts. I'm hoping in some cases readability can still be improved, but I'd also have to read up in more detail about class groups to fully review them.

DemiMarie committed:
Fix copyright notices and add some docs to ffi.rs

The copyright notices listed “Block Notary Inc” as a copyright owner,
which is wrong.  The correct owner is POA Networks, Ltd.

Also, add some doc comments to ffi.rs.

DemiMarie committed:
Add useful scripts

- `compare.sh` runs the CLI and `pot` and compares the results
- `format_logs.awk` formats log produced by my modified version of pot
  into a form that can be used as test cases.
- `strip_extra_whitespace.sh` strips trailing whitespace, as well as
  leading and trailing lines containing only whitespace, from every file
  in the repository.  It also runs `cargo fmt`.
- `test.sh` just runs repeated squaring with the parameters provided as
  an example for the C++ example submission to the VDF competition.

DemiMarie committed:
Run `cargo fmt`.

mbr commented on vdf/build.rs:
```diff
@@ -0,0 +1,114 @@
+    res
+}
+
+// Probably horribly slow.  I don’t care.
```
Probably good enough: https://github.com/poanetwork/threshold_crypto/blob/4fec9da3d6207a89f82ec8137f30cd8cfc9c7081/src/mock/ms8.rs#L19

DemiMarie committed:
Document I/O errors in test cases

DemiMarie committed:
Document build.rs

DemiMarie committed:
Avoid `#[forbid(warnings)]`

Sometimes, warnings are unavoidable, such as in automatically-generated
code or when clippy has false positives.

DemiMarie committed:
Fix the strip-trailing-whitespace script

The sed script was wrong, and it shouldn’t automatically run `cargo
build`.

DemiMarie committed:
Clarify the `Classgroup` trait

especially the requirements on
`ClassGroup::unsigned_deserialize_bignum`.

DemiMarie committed:
Typo: indeces → indices

DemiMarie committed:
Use vec![value; length] where possible

instead of Vec::with_capacity() and Vec::resize()

DemiMarie committed:
Use `std::borrow::Borrow` to clean up trait impls

This avoid redundantly implementing traits twice.

afck commented on vdf/build.rs:
```diff
@@ -0,0 +1,114 @@
+            }
+        }
+        residues.push(x);
+    }
```
More idiomatic (not sure if more readable):

```rust
let residues: Vec<usize> = {
    let primes = [3, 5, 7, 11, 13];
    let not_divisible = |x: usize| primes.iter().all(|p| x % p != 0);
    (7..M).step_by(8).filter(not_divisible).collect()
};
```

DemiMarie replied:
I changed this loop, as well as some others, to operate over iterators.  I could also merge the production of the list of prime numbers with its consumption, but I figured that the code as-is is more readable.

afck reviewed (commented):
In the places where I asked for documentation and where we're still unsure ourselves, we can also just add a `TODO` and point to the C++ or Python code for now, and that we want to be compatible with it.

DemiMarie committed:
Clean up discriminant generation tests

Split one test into two, and reformat.

DemiMarie committed:
Document create_discriminant(), ditch transmute

Also replace “entropy” with “random_bytes” in identifiers.

DemiMarie committed:
Clean up discriminant creation

The old code was ugly and difficult to read.

DemiMarie committed:
Remove premature optimization in main.rs

It is okay to allocate a byte array twice.

DemiMarie committed:
Remove unused imports in vdf/tests/check_multiply.rs

DemiMarie committed:
Make argument to `odd_primes_below_n()` a constant

The function is only designed to be used with a single specific
argument, so make that argument a constant and the function into one
with no arguments.

DemiMarie committed:
Remove redundant assertions

They were checking the validity of `GmpClassGroup`s that were already
being checked for validity elsewhere, without any intervening
modifications to those `GmpClassGroup`s.

DemiMarie committed:
Use iterators to clean up build script

The old code was quite unideomatic.

DemiMarie committed:
Document `CongruenceContext` and its methods

DemiMarie committed:
Basic constant propagation in `GmpClassGroup::multiply`

In `GmpClassGroup::multiply`, `ctx.r` was set to zero and never set
afterwards.  Therefore, any operations it is in can be constant-folded.
This also opens more opportunities for constant-folding in this
function.

DemiMarie committed:
Delete dead code

DemiMarie committed:
Shrink `CongruenceContext` by 1 member

DemiMarie committed:
Run `cargo fix --edition` to prepare for Rust 2018

DemiMarie committed:
Document the poorly-named `import_obj` and `export_obj`

Also do some minor refactoring.

DemiMarie committed:
Add documentation of prime number generation.

afck commented on vdf/build.rs:
```diff
@@ -16,12 +16,22 @@ use std::fs::File;
 use std::path::PathBuf;
 use std::u16;
+
+/// The number of odd prime less than 65536.
```
Typo: "primes"

DemiMarie replied:
Fixed.

afck commented on vdf/src/create_discriminant.rs:
```diff
@@ -58,26 +55,31 @@ fn random_bytes_from_seed(seed: &[u8], byte_count: usize) -> Vec<u8> {
+    n.setbit(usize::from(length - 1));
+    n = -n;
+    let rem = mpz_rem_u32(&n, M);
+    n = -n;
```
Is there a reason not just to

```rust
let rem = mpz_rem_u32(&(-n), M);
```

and avoid reassigning `n` twice?

DemiMarie replied:
I suspect that doing it this way avoids a copy.

DemiMarie replied:
Also, `mpz_rem_u32` is a (very) poor name ― it calls `__gmpz_cdiv_ui`, rather than `__gmpz_tdiv_ui` as would be expected.

DemiMarie replied:
To elaborate: GMP uses sign-magnitude internally, so flipping the sign of a number is cheap.  Furthermore, `n = -n` actually modifies `n` , rather than needing to copy it.

afck commented on vdf/src/gmp_classgroup/congruence.rs:
```diff
@@ -36,6 +45,13 @@ impl Default for CongruenceContext {
+///
+/// # Panics
+///
+/// Panics if the congruence could not be solved.
```
Should that comment be on `solve_linear_congruence` instead of `CongruenceContext`?

DemiMarie replied:
Fixed

afck reviewed (commented):
The changes look great so far!

DemiMarie committed:
Clean up remainders in discriminant creation

The code had used ceiling division to calculate remainders, and changed
the sign of the numerator where required.  This was confusing and made
the code bug-prone.  Instead, use floor division where it is needed and
ceiling division where it is needed.  This makes the code clearer and
easier to read.

DemiMarie committed:
Add `rustfmt.toml` and `LICENSE_TEMPLATE`

and run `cargo-fmt`.

DemiMarie committed:
Fix typos

`prime` → `primes` in `build.rs`, and fix a misplaced doc comment in
`congruence.rs`

DemiMarie committed:
Tell rustfmt to be more aggressive; update .gitignore

DemiMarie committed:
More reformatting

DemiMarie committed:
Add test harness

This adds 468 different test cases, one for each length between 66 and
1000 inclusive.  It also adds a shell script to test them with.  All
tests pass.

DemiMarie committed:
Document the difference between a generator and the identity

Also rename `generate_for_discriminant` to `generator_for_discriminant`.

DemiMarie committed:
Clean up CLI and API

This adds an `Iterations` type for the number of iterations.  This is a
newtype of `u64`, but guarantees that it contains a valid number of
iterations.  It also cleans up the command-line interface, reducing its
amount of code.

DemiMarie committed:
Replace long if-else chain by pattern matching

DemiMarie committed:
Add a comment to the constant `M` in build.rs

DemiMarie committed:
Fix some clippy warnings

DemiMarie committed:
Add links to the inkfish tool’s code

DemiMarie committed:
Document purpose of `approximate_i`.

DemiMarie committed:
Add additional tests to check.sh and rename it to ci.sh

DemiMarie committed:
`test.sh` should use `cargo run`

DemiMarie committed:
Fix typo: “POA Networks, Ltd.” → “POA Networks Ltd.”

DemiMarie committed:
Wesolowski proof-of-time proofs implemented & tested

Both the proof and verification work and pass tests.

DemiMarie committed:
SECURITY FIX: fix primality checking in Wesolowski proof

Use 25 Miller-Rabin iterations, not 2.

DemiMarie commented:
Both types of proofs are working.  Verification is also working.

c0gent commented:
Demi this PR has become far too large, probably because we haven't been keeping up with your work and you haven't been able to commit your changes directly.

If @afck and everyone else agrees I think we should rebase-merge this now, as-is, and make any needed corrections later.

DemiMarie commented:
@c0gent I don’t think rebasing is necessary.  There should be no commits in `master` that are not in this branch.

c0gent commented:
No I just meant that I didn't want a merge commit. In other words it will fast forward merge.

DemiMarie committed:
Mega refactoring: Encapsulate VDF internals

This commit encapsulates the VDF internals in a pair of traits, `VDF`
and `VDFParams`.  `VDF` is object-safe, which is quite useful in some
contexts.  The `ClassGroup` code is made private, and will be made into
a separate crate later.

DemiMarie committed:
Update to Rust 2018 edition

These are not exactly the changes suggested by `cargo fix`, though.

DemiMarie committed:
Remove GMP dependencies from classgroup code

This is necessary if we wish to provide other backends, such as ones
based on OpenSSL or TomsFastMath.

DemiMarie committed:
Improve documentation

Also, use a consistent message when internal serialization fails due to
an incorrect buffer size calculation.

DemiMarie committed:
Delete dead code in classgroup.rs

DemiMarie committed:
Create new `classgroup` crate

This holds the class group implementation and related traits.

Also, fix `rustfmt.toml` to work on stable Rust.

DemiMarie committed:
Add README.md and `vdf` crate-leve doc comment

c0gent commented:
Demi, go ahead and do a "Rebase and merge" with this PR. You can then either push commits to master directly or create a PR if you want help or a review.
